### ISSUE TITLE
False positive: Callable arg types more strict than required

### ISSUE THREAD

**Issue by Phil-Barber** [python/mypy#15328]
Struggled to name this properly but here's a repro:

```
import typing

# Want to accept any callable with a string arg
CallableT = typing.Callable[[str], None]

def func(callable_m: dict[str, CallableT]) -> None:
    callable_m['first']("a")
    
    
def first(a: str) -> None:
    return None
    
def second(a: str) -> None:
    return None

def third(b: str) -> None:
    return None   

# This mapping will fail because the callables will have type:
# Callable[[Arg(str, 'a')], None]    
failed_mapping = {
    'first': first,
    'second': second
}
func(failed_mapping)

# By renaming the arg in `third` from `a` to `b` everything is groovy
passing_mapping = {
    'first': first,
    'third': third
}
func(passing_mapping)
```

https://mypy-play.net/?mypy=latest&python=3.11&gist=ff039c923e370f79a680c463ef627810

Easy workaround with renaming the arg but is a bit of a pain

**Comment by erictraut** [comment]
Another way to work around this bug is to mark the parameters as positional-only. This is probably the same amount of work as renaming the parameters, but I figured I'd mention it in case you find it more palatable.

```python
def second(a: str, /) -> None:
    return None
```
